File: mobile app/lib/core/utils/upi_app_launcher.dart
import 'package:flutter/material.dart';
import 'package:url_launcher/url_launcher.dart';

import 'upi_payment.dart';

class UpiAppOption {
  const UpiAppOption({
    required this.id,
    required this.label,
    required this.shortLabel,
    required this.color,
    required this.uriPrefix,
  });

  final String id;
  final String label;
  final String shortLabel;
  final Color color;
  final String uriPrefix;

  String buildUri(String query) => '$uriPrefix$query';
}

class UpiAppLauncher {
  UpiAppLauncher._();

  static const options = <UpiAppOption>[
    UpiAppOption(
      id: 'gpay',
      label: 'Google Pay',
      shortLabel: 'GPay',
      color: Color(0xFF4285F4),
      uriPrefix: 'tez://upi/pay?',
    ),
    UpiAppOption(
      id: 'phonepe',
      label: 'PhonePe',
      shortLabel: 'Pe',
      color: Color(0xFF5F259F),
      uriPrefix: 'phonepe://pay?',
    ),
    UpiAppOption(
      id: 'paytm',
      label: 'Paytm',
      shortLabel: 'Tm',
      color: Color(0xFF00BAF2),
      uriPrefix: 'paytmmp://pay?',
    ),
    UpiAppOption(
      id: 'bhim',
      label: 'BHIM UPI',
      shortLabel: 'BHIM',
      color: Color(0xFF0984E3),
      uriPrefix: 'bhim://pay?',
    ),
    UpiAppOption(
      id: 'generic',
      label: 'Other UPI',
      shortLabel: 'UPI',
      color: Color(0xFF6366F1),
      uriPrefix: 'upi://pay?',
    ),
  ];

  static Future<List<UpiAppOption>> detectInstalledApps() async {
    final available = <UpiAppOption>[];
    final primaryApps = options.where((option) => option.id != 'generic');

    for (final option in primaryApps) {
      final canOpen = await canLaunchUrl(Uri.parse('${option.uriPrefix}pa=test@upi'));
      if (canOpen) available.add(option);
    }

    if (available.isNotEmpty) return available;

    // Fallback: show popular UPI apps — installed ones open on tap.
    return primaryApps.toList();
  }

  static Future<bool> launchApp({
    required UpiAppOption app,
    required double amount,
    required String note,
    String? merchantUpiId,
    String? merchantUpiName,
  }) async {
    final query = UpiPayment.buildPaymentQuery(
      amount,
      note,
      merchantUpiId: merchantUpiId,
      merchantUpiName: merchantUpiName,
    );
    if (query.isEmpty) return false;

    final uri = Uri.parse(app.buildUri(query));
    return launchUrl(uri, mode: LaunchMode.externalApplication);
  }
}

File: frontend/src/utils/upiAppLauncher.js
import { buildUpiQuery, isMobileDevice } from "./upiPayment";

export const UPI_APP_OPTIONS = [
  {
    id: "gpay",
    label: "Google Pay",
    shortLabel: "GPay",
    color: "#4285F4",
    uriPrefix: "tez://upi/pay?",
  },
  {
    id: "phonepe",
    label: "PhonePe",
    shortLabel: "Pe",
    color: "#5F259F",
    uriPrefix: "phonepe://pay?",
  },
  {
    id: "paytm",
    label: "Paytm",
    shortLabel: "Tm",
    color: "#00BAF2",
    uriPrefix: "paytmmp://pay?",
  },
  {
    id: "bhim",
    label: "BHIM UPI",
    shortLabel: "BHIM",
    color: "#0984E3",
    uriPrefix: "bhim://pay?",
  },
];

export function getUpiAppOptions() {
  return UPI_APP_OPTIONS;
}

export function buildUpiAppUrl(app, amount, note, config = {}) {
  const query = buildUpiQuery(amount, note, config);
  if (!query) return "";
  return `${app.uriPrefix}${query}`;
}

export function openUpiApp(app, amount, note, config = {}) {
  const query = buildUpiQuery(amount, note, config);
  if (!query) return false;

  const ua = navigator.userAgent || "";
  const isAndroid = /android/i.test(ua);
  const appUrl = `${app.uriPrefix}${query}`;

  let targetUrl = appUrl;
  if (isAndroid && app.id === "generic") {
    targetUrl = `intent://pay?${query}#Intent;scheme=upi;end`;
  } else if (isAndroid) {
    targetUrl = appUrl;
  }

  const link = document.createElement("a");
  link.href = targetUrl;
  link.rel = "noopener noreferrer";
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);

  return isMobileDevice();
}
